lisa/operating_system.py

```diff
@@ -225,11 +409,68 @@ class Fedora(Linux):
     def name_pattern(cls) -> Pattern[str]:
         return re.compile("^Fedora|fedora$")
 
-    def _install_packages(self, packages: Union[List[str]]) -> None:
-        self._node.execute(
-            f"dnf install -y {' '.join(packages)}",
-            sudo=True,
+    def _install_packages(
+        self, packages: Union[List[str]], signed: bool = True
+    ) -> None:
+        command = f"dnf install -y {' '.join(packages)}"
+        if not signed:
+            command.__add__(" --nogpgcheck")
+
+        install_result = self._node.execute(command, sudo=True)
+        if install_result.exit_code != 0:
+            raise LisaException(
+                f"Failed to install {' '.join(packages)}."
+                f" stdout: {install_result.stdout}"
+            )
+        else:
+            self._log.info(
+                f"{' '.join(packages)} is/are installed successfully."
+                f" stdout: {install_result.stdout}"
+            )
+
+    def _package_exists(self, package: str, signed: bool = True) -> bool:
+        command = f"dnf list installed | grep -Ei {package}"
+        result = self._node.execute(command, sudo=True)
+        if result.exit_code == 0:
+            for row in result.stdout.split("\n"):
+                if package in row:
+                    return True
+
+        return False
+
+    def _get_os_version(self) -> OsVersion:
+        if self._os_version:
+            return self._os_version
+
+        self._os_version = OsVersion("")
+        cmd_result = self._node.execute(
+            # Typical output of 'cat /etc/fedora-release' is -
+            # Fedora release 22 (Twenty Two)
+            cmd="cat /etc/fedora-release",
+            no_error_log=True,
         )
+        if cmd_result.exit_code == 0 and cmd_result.stdout != "":
+            result = cmd_result.stdout
+            for vendor in ["Fedora", "CentOS", "Red Hat", "XenServer"]:
```

Comment:
This is in Ubuntu class, why need to check those strings?

Reply:
This is Fedora class. Ubuntu is not even implementing _get_os_version(), it is deriving from Debian

Reply:
I insist that ifs without elses should test for the negative, instead, and pass/continue if it does not apply, leaving the actual logic one level less indented, and easier to read, but it seems I'm the only one thinking like that here. Also, why have a default _os_version empty and not null instead?